tuta-sdk/rust/src/crypto/entity_facade.rs

```diff
@@ -25,38 +25,42 @@ impl EntityFacade {
+
+        for (key, model_value) in type_model.values.iter() {
+            let stored_element = entity.remove_entry(key).unwrap_or_else(|| (key.to_string(), ElementValue::Null)).1;
+            let element_value = self.map_value(stored_element, session_key, key, model_value);
             match element_value {
                 Ok((decrypted, ivs, errors)) => {
                     mapped_decrypted.insert(key.to_string(), decrypted);
-                    mapped_decrypted.get("errors").unwrap().assert_dict().insert(key.to_string(), ElementValue::Dict(errors));
-                    mapped_decrypted.get("final_ivs").unwrap().assert_dict().insert(key.to_string(), ElementValue::Dict(ivs))
+                    mapped_errors.insert(key.to_string(), ElementValue::Dict(errors));
+                    mapped_ivs.insert(key.to_string(), ivs)
                 }
                 Err(err) => return Err(err)
             };
         }
 
-        let model_name = &type_model.name;
-        for (association_name, association_model) in type_model.associations.clone().into_iter() {
-            let association = self.map_associations(type_model, entity.to_owned(), session_key, model_name, &association_name, association_model);
+        for (association_name, association_model) in type_model.associations.iter() {
+            let (_, association_entry) = entity.remove_entry(association_name).unwrap_or((association_name.to_string(), ElementValue::Null));
+            let association = self.map_associations(type_model, association_entry, session_key, &association_name, association_model.to_owned());
             match association {
                 Ok((mapped_association, errors)) => {
                     mapped_decrypted.insert(association_name.to_string(), mapped_association);
-                    mapped_decrypted.get("errors").unwrap().assert_dict().insert(association_name.to_string(), ElementValue::Dict(errors));
+                    mapped_errors.insert(association_name.to_string(), ElementValue::Dict(errors));
                 }
                 Err(err) => return Err(err)
             }
         }
 
+        mapped_decrypted.insert("errors".to_string(), ElementValue::Dict(mapped_errors));
+        mapped_decrypted.insert("final_ivs".to_string(), ElementValue::Dict(mapped_ivs));
         Ok(mapped_decrypted)
     }
 
-    fn map_associations(&self, type_model: &TypeModel, entity: ParsedEntity, session_key: &AesKey, model_name: &String, association_name: &String, association_model: ModelAssociation) -> Result<(ElementValue, HashMap<String, ElementValue>), ApiCallError> {
+    fn map_associations(&self, type_model: &TypeModel, association_data: ElementValue, session_key: &AesKey, association_name: &String, association_model: ModelAssociation) -> Result<(ElementValue, HashMap<String, ElementValue>), ApiCallError> {
         let mut errors: HashMap<String, ElementValue> = Default::default();
         let dependency = match association_model.dependency {
```

Comment:
```suggestion
        let dependency = match association_model.dependency {
            Some(ref dep) => dep,
            None => &type_model.app
        };
```

if you do this you don't have to move `association_model`, you can borrow it